File: sphone-api/Domain/Customers/Customer.cs
using SPhone.Domain.Common;
using SPhone.Domain.Customers.Events;

namespace SPhone.Domain.Customers;

public sealed class Customer : AggregateRoot<CustomerId>
{
    public NationalId NationalId { get; private set; }
    public string FullName { get; private set; }
    public PhoneNumber PhoneNumber { get; private set; }
    public DateOnly DateOfBirth { get; private set; }
    public decimal CreditLimit { get; private set; }
    public CustomerStatus Status { get; private set; }
    public DateTime CreatedAt { get; private set; }

    // EF Core constructor
    private Customer() : base(CustomerId.New())
    {
        NationalId = null!;
        FullName = null!;
        PhoneNumber = null!;
    }

    private Customer(
        CustomerId id,
        NationalId nationalId,
        string fullName,
        PhoneNumber phoneNumber,
        DateOnly dateOfBirth) : base(id)
    {
        NationalId = nationalId;
        FullName = fullName;
        PhoneNumber = phoneNumber;
        DateOfBirth = dateOfBirth;
        CreditLimit = 30_000m; // Default credit limit 30,000 THB
        Status = CustomerStatus.Pending;
        CreatedAt = DateTime.UtcNow;
    }

    public static Customer Register(
        NationalId nationalId,
        string fullName,
        PhoneNumber phoneNumber,
        DateOnly dateOfBirth)
    {
        if (string.IsNullOrWhiteSpace(fullName))
            throw new DomainException("Full name is required.");

        var customer = new Customer(CustomerId.New(), nationalId, fullName, phoneNumber, dateOfBirth);
        customer.RaiseDomainEvent(CustomerRegisteredEvent.Create(customer.Id, nationalId, fullName, phoneNumber));
        return customer;
    }

    public void Verify()
    {
        if (Status != CustomerStatus.Pending)
            throw new DomainException("Only pending customers can be verified.");
        Status = CustomerStatus.Verified;
        CreditLimit = 50_000m; // Increase limit upon verification
    }

    public void Suspend()
    {
        if (Status == CustomerStatus.Suspended)
            throw new DomainException("Customer is already suspended.");
        Status = CustomerStatus.Suspended;
    }
}

File: sphone-api/src/domain/customers/Customer.ts
import { AggregateRoot } from '../shared/AggregateRoot'
import { CustomerId } from './CustomerId'
import { NationalId } from './NationalId'
import { PhoneNumber } from './PhoneNumber'
import { CustomerStatus } from './CustomerStatus'
import { CustomerRegisteredEvent } from './events/CustomerRegisteredEvent'

export class Customer extends AggregateRoot {
  private constructor(
    private readonly _id: CustomerId,
    private readonly _nationalId: NationalId,
    private _fullName: string,
    private readonly _phoneNumber: PhoneNumber,
    private readonly _dateOfBirth: Date,
    private _creditLimit: number,
    private _status: CustomerStatus,
    private readonly _createdAt: Date,
  ) {
    super()
  }

  get id(): CustomerId { return this._id }
  get nationalId(): NationalId { return this._nationalId }
  get fullName(): string { return this._fullName }
  get phoneNumber(): PhoneNumber { return this._phoneNumber }
  get dateOfBirth(): Date { return this._dateOfBirth }
  get creditLimit(): number { return this._creditLimit }
  get status(): CustomerStatus { return this._status }
  get createdAt(): Date { return this._createdAt }

  static create(params: {
    id: string
    nationalId: string
    fullName: string
    phoneNumber: string
    dateOfBirth: Date
  }): Customer {
    const customer = new Customer(
      CustomerId.create(params.id),
      NationalId.create(params.nationalId),
      params.fullName,
      PhoneNumber.create(params.phoneNumber),
      params.dateOfBirth,
      100000,
      CustomerStatus.Pending,
      new Date(),
    )
    customer.addDomainEvent(
      new CustomerRegisteredEvent(params.id, params.nationalId, params.fullName),
    )
    return customer
  }

  static reconstitute(params: {
    id: string
    nationalId: string
    fullName: string
    phoneNumber: string
    dateOfBirth: Date
    creditLimit: number
    status: number
    createdAt: Date
  }): Customer {
    return new Customer(
      CustomerId.create(params.id),
      NationalId.create(params.nationalId),
      params.fullName,
      PhoneNumber.create(params.phoneNumber),
      params.dateOfBirth,
      params.creditLimit,
      params.status as CustomerStatus,
      params.createdAt,
    )
  }

  canPlaceOrder(): boolean {
    return this._status === CustomerStatus.Pending || this._status === CustomerStatus.Verified
  }
}
